polyguice-config/src/main/java/com/flipkart/polyguice/config/ApacheCommonsConfigProvider.java

```diff
@@ -45,36 +48,25 @@ public ApacheCommonsConfigProvider() {
         rootConfig = new CompositeConfiguration();
     }
 
-    public ApacheCommonsConfigProvider location(String loc) {
+    public ApacheCommonsConfigProvider location(String filePath) {
+        Class<? extends Configuration> clazz = selectConfigImpl(filePath);
         try {
-            if(loc.toLowerCase(Locale.getDefault()).endsWith(".properties")) {
-                PropertiesConfiguration config = new PropertiesConfiguration(new File(loc));
-                rootConfig.addConfiguration(config);
-                LOGGER.debug("properties configuration from {}", loc);
-            }
-            else if(loc.toLowerCase(Locale.getDefault()).endsWith(".xml")) {
-                XMLConfiguration config = new XMLConfiguration(new File(loc));
-                rootConfig.addConfiguration(config);
-                LOGGER.debug("xml configuration from {}", loc);
-            }
-            else if(loc.toLowerCase(Locale.getDefault()).endsWith(".json")) {
-                JsonConfiguration config = new JsonConfiguration(new File(loc));
-                rootConfig.addConfiguration(config);
-                LOGGER.debug("json configuration from {}", loc);
-            }
-            else if(loc.toLowerCase(Locale.getDefault()).endsWith(".yml")) {
-                YamlConfiguration config = new YamlConfiguration(loc);
-                rootConfig.addConfiguration(config);
-                LOGGER.debug("yaml configuration from {}", loc);
-            }
-            else if(loc.toLowerCase(Locale.getDefault()).endsWith(".yaml")) {
-                YamlConfiguration config = new YamlConfiguration(loc);
-                rootConfig.addConfiguration(config);
-                LOGGER.debug("yaml configuration from {}", loc);
+            if (clazz != null) {
+                this.rootConfig.addConfiguration(clazz.getConstructor(File.class).newInstance(new File(filePath)));
             }
+        } catch (InstantiationException | IllegalAccessException | NoSuchMethodException | InvocationTargetException e) {
+            throw new RuntimeException(e);
         }
-        catch(Exception exep) {
-            LOGGER.error("unable to load configuration from " + loc.toString(), exep);
+        return this;
+    }
+    public ApacheCommonsConfigProvider location(URL url) {
```

Comment:
Shouldn't it be the other way round? Take only URLs - the standard way of identifying resources. By identifying the "scheme" part of a given string, we essentially end up re-writing the logic that is contained within URL.

In fact, I kept the String path location to maintain backward compatibility. 

Also, if we give something like file:/<some-jar>.jar!/foo.yaml ; the approach you mentioned fails. Although the scheme says "file", it is not a file in the file system but a file contained within a jar. We can possibly work around this by providing classpath:/foo.yaml. However, the former URL is the one that is provided when we do a classLoader.getResource(). Identifying a resource using the classLoader seems more straight forward and less error prone than constructing the string "classpath:/foo.yaml" and then passing it to the config provider.